include/envoy/http/header_map.h

```diff
@@ -500,6 +517,15 @@ class HeaderMap {
    */
   virtual void setCopy(const LowerCaseString& key, absl::string_view value) PURE;
 
+  /**
+   * Saves the original casing of header key in the map alongside its lower-case form.
+   * The key MUST point to point to data that will live beyond the lifetime of any
+   * request/response using the string (since a codec may optimize for zero copy).
+   *
+   * @param key specifies the name of the header to set
+   */
+  virtual void preserveCase(const LowerCaseString& key) PURE;
```

Comment:
Can we avoid having this call? Instead have the case preservation be an option on key/value insert...I don't think we want to store the original key on LowerCaseString. Being on the HeaderEntry should be enough.